### PR TITLE
Change the working directory if we can't stat it 

### LINKED ISSUE #4634: Cannot create a new Gitea user without Gitea system user having read/execute access to current directory
- Gitea version (or commit ref): 1.4.3
- Git version: 2.18.0
- Operating system: FreeBSD 11.2
- Database (use `[x]`):
  - [ ] PostgreSQL
  - [ ] MySQL
  - [ ] MSSQL
  - [x] SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - [ ] Yes (provide example URL)
  - [ ] No
  - [x] Not relevant
- Log gist:

## Description
Trying to create a new user using command-line tools (`sudo -u git gitea admin create-user --name=user --password=pass --email=user@example.com --config=/usr/local/etc/gitea/conf/app.ini`, git being Gitea user) while being in a directory without Gitea user having access to it (like `/home/john` with `700` permissions) causes this error:
```
panic: error getting work directory: stat .: permission denied                                                                                                                    
                                                                                                                                                                                  
goroutine 1 [running]:                                                                                                                                                            
code.gitea.io/gitea/vendor/gopkg.in/macaron%2ev1.init.1()                                                                                                                         
        /wrkdirs/usr/ports/www/gitea/work/src/code.gitea.io/gitea/vendor/gopkg.in/macaron.v1/macaron.go:317 +0x125
```
Entering a different directory with git having access to it (`/` for example) fixes this issue. Honestly I do not really understand why gitea even checks the current directory, as it has nothing to do with what it is doing.

### PR DESCRIPTION
**Pull request by zeripath**
Attempt to change the working directory if we can't stat it. This hopefully runs before the macaron init but may need to be moved to a file called a.go

Fix #4634

### CODE REVIEW

**Comment by zeripath** [comment]
OK this is now working but it's not excellent as it doesn't inform us that we are moving a temporary directory. This needs to be logged somehow but will require thought - for example - we can't log from gitea serv as this would break git.

**Comment by mrsdizzie** [comment]
Does it make sense for this code to run all of the time? If somebody runs ```gitea web``` in a folder they don't own, it wouldn't make sense to move to a /tmp folder and create files/run from there would it? Or to create a dump file in a /tmp folder and then delete that /tmp folder after run

If not, could it just be run based on os.Args and only run for certain commands where using a temporary work dir makes sense, like ```gitea admin```, and avoiding other issues like the one you mention? 

I assume it isn't reasonable to have gitea configured to run from a place without proper permissions and have things like web and srv working, where as the initial bug report case is more reasonable to expect

**Comment by zeripath** [comment]
@mrsdizzie You don't get that choice - this has to run at package init() time, and has to run before macaron.v1's package init runs.

That means when it runs you don't know what command you're going to run.

**Comment by zeripath** [comment]
We could however, detect that we're running from the temporary directory for some commands and fail because we can't run from there with a different Before for those that need it.

File: aainit/init.go [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+	if runtime.GOOS == "windows" {
+		OriginalWorkingDirectory = os.Getenv("CD")
+	} else {
+		OriginalWorkingDirectory = os.Getenv("PWD")
+	}
+	wd, err := os.Getwd()
+	if err == nil {
+		OriginalWorkingDirectory = wd
+		return
+	}
```
Comment (lunny) [inline aainit/init.go]: Maybe we should print the error reason.

Comment (zeripath) [reply aainit/init.go]: OK done.

**Comment by zeripath** [comment]
An alternative option is to delete the working directory at the start of main. Then if anything tries to write or read from the working directory it should fail.

**Pull request closed without merging**